### ISSUE TITLE
Cross Platform Keyboard Input

### ISSUE THREAD

**Issue by belak** [cocos2d/cocos2d-x#2485]
I may be missing something, but I've been trying for quite a while to get keyboard input for controls that works on a cross platform basis.

If there isn't something already, it makes sense to have something that can handle input easily.

EDIT: this is one of the few things I feel cocos2d-x is extremely lacking in. When developing for the PC, if you only have touches easily available, it is extremely limiting.

**Comment by javienen** [comment]
+1, I saw that someone made an OSx keyboard implementation but I don't know why it hasn't been added on the main trunk, and there isn't windows keyboard support either. I did a simple implementation, but It would be great to be included on the source of the next release of cocos 2d-x.

Cocos 2d-x is awesome by the way, it just needs some adjustments and I could be more popular =D.

**Comment by belak** [comment]
As far as I can tell it's "possible" by using a CCEGLView but I haven't been able to get it to work and it seems like a really hacky solution. Once this is added to cocos2d-x, I feel it would work great as desktop development platform as well as mobile.

**Comment by belak** [comment]
Ok, I've been looking into this. It honestly shouldn't be a huge issue, but I don't know how the codebase works.

What needs to be done:
- Keyboard support needs to be added to related classes - this was mostly done in the patch to add keyboard support for Macs
- CCEGLView needs to be modified on at least Windows and Linux to allow for keyboard input (see http://asmak9.blogspot.com/2012/08/cocos2d-x-enable-hardware-keyboard.html, as well as looking into modifying the keyboard callbacks for glfw on Linux - which are already in place)
- Some form of matching key codes needs to happen (if they don't just use ascii), possibly a translation of most out of the ordinary keycodes to a cocos format. 

The largest part is ripping out the keyboard structure from the previous commits, but it should still be possible.

**Comment by reneklacan** [comment]
https://github.com/cocos2d/cocos2d-x/pull/2704 this merged PR fixed this issue

**Comment by belak** [comment]
Sounds like Windows support is still missing. So this isn't quite fixed but it's a huge step forward.

**Comment by reneklacan** [comment]
just define KEYBOARD_SUPPORT directive and it should work just ok on windows too :)

**Comment by belak** [comment]
My mistake. I only saw Mac and Linux mentioned in the PR. I'll check it out.

**Comment by reneklacan** [comment]
I'm not 100% sure but from design there should be no problems (maybe you will have to add KeyboardDispatcher.cpp to the list of files to compile

**Comment by belak** [comment]
The changes look great for Linux and OSX, but I don't see any changes in the cocos2dx/platform/win32/ folder.

Also, I'm not sure if it's worth supporting keyboards/docks for Android/iOS, but that may be something to look into for the future.

EDIT: That being said, I'm really looking forward to playing with this. It's been essentially the last thing holding me back from using cocos2d-x as a platform.

**Comment by reneklacan** [comment]
Oh, I forgot about that, sorry ... I will try to add that code next time I will be on some windows machine :)

**Comment by cobafesbuk** [comment]
Hi, I've just finished trying some game engines and I found cocos2d-x is the most capable but the problem is when I'm searching for keyboard listener, I got result stuck with results as hacking the engine which I don't really understand how to.
I just want to learn how to make a simple game but same as belak, the keyboard support draw me back, luckily this month you make the patch for it :)

I found out about your patch here
http://www.cocos2d-x.org/boards/6/topics/28973?r=29622

I'm really new to C++, VS, and repository things. I don't even know how to enable your patch to the cocos. I only just finished the making a helloworld :D

I'm using cocos2d-2.1rc0-x-2.1 now.
Thanks for the patch :)
